<a href="https://colab.research.google.com/github/midhun-james/gliner-desc/blob/main/rag_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install -q -U google-generativeai langchain-google-genai langchain-community chromadb sentence-transformers

In [2]:
pip install -q pymupdf

In [3]:
import os
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = ''
os.environ['NO_GCE_CHECK'] = 'True'

Setting up the api key configuration

In [29]:
from google.colab import userdata
import os

os.environ["GOOGLE_API_KEY"] = userdata.get('GOOGLE_API_KEY')

llm = ChatGoogleGenerativeAI(
    model="gemini-2.5-flash",
    temperature=0,
    max_retries=2
)


In [13]:
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.vectorstores import Chroma
from langchain_community.embeddings import SentenceTransformerEmbeddings
from langchain.chains import RetrievalQA
from langchain_google_genai import ChatGoogleGenerativeAI

Copied the text in 10-k reports to a txt file.

Used fitz package from Pymupdf to extract text from the pdf

In [12]:
import fitz
pdf_path='assign_3_data.txt'
doc=fitz.open(pdf_path)
text=''
for page in doc:
  text+=page.get_text()
print(text[:500])

 
 
UNITED STATES
SECURITIES AND EXCHANGE COMMISSION
Washington, D.C. 20549
 
FORM 10-K
 
 
☒
ANNUAL REPORT PURSUANT TO SECTION 13 OR 15(d) OF THE
SECURITIES EXCHANGE ACT OF 1934
 
 
 
For the Fiscal Year Ended June 30, 2022
 
 
 
OR
 
 
☐
TRANSITION REPORT PURSUANT TO SECTION 13 OR 15(d) OF
THE SECURITIES EXCHANGE ACT OF 1934
 
 
 
For the Transition Period From                  to
Commission File Number 001-37845
 
MICROSOFT CORPORATION
 
 
Washington
 
91-1144442
(STATE OF INCORPORATION)
 
(I


For text splitting setted the chunk size as 1500 and chunk overlap as 200.

Since it is a very large file chunk making chunk size bigger is needed. overlapping is setted inorder to not loose continuity

In [14]:
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=1500,
    chunk_overlap=200,
    separators=["\n\n", "\n", ".", " ", ""]
)
chunks=text_splitter.split_text(text)



In [15]:
print(f'Total chunks: {len(chunks)}')
print(f'first chunk : {chunks[0]}')

Total chunks: 286
first chunk : UNITED STATES
SECURITIES AND EXCHANGE COMMISSION
Washington, D.C. 20549
 
FORM 10-K
 
 
☒
ANNUAL REPORT PURSUANT TO SECTION 13 OR 15(d) OF THE
SECURITIES EXCHANGE ACT OF 1934
 
 
 
For the Fiscal Year Ended June 30, 2022
 
 
 
OR
 
 
☐
TRANSITION REPORT PURSUANT TO SECTION 13 OR 15(d) OF
THE SECURITIES EXCHANGE ACT OF 1934
 
 
 
For the Transition Period From                  to
Commission File Number 001-37845
 
MICROSOFT CORPORATION
 
 
Washington
 
91-1144442
(STATE OF INCORPORATION)
 
(I.R.S. ID)
 
ONE MICROSOFT WAY, REDMOND, Washington 98052-6399
(425) 882-8080
www.microsoft.com/investor
 
 
 
 
 
Securities registered pursuant to Section 12(b) of
the Act:
 
 
 
 
 
 
 
 
 
Title of each class
 
Trading Symbol
 
Name of exchange on which registered
 
 
 
 
 
Common stock, $0.00000625 par value per share
 
MSFT
 
Nasdaq
3.125% Notes due 2028
 
MSFT
 
Nasdaq
2.625% Notes due 2033
 
MSFT
 
Nasdaq
 
 
 
 
 
Securities registered pursuant to Section 12(g

For RAG we need these text chunks as documents.

By doing this we can get a metadata along with the text content

In [16]:
docs = text_splitter.create_documents([text])
print(f"Total documents: {len(docs)}")
print(docs[0].page_content[:300])

Total documents: 286
UNITED STATES
SECURITIES AND EXCHANGE COMMISSION
Washington, D.C. 20549
 
FORM 10-K
 
 
☒
ANNUAL REPORT PURSUANT TO SECTION 13 OR 15(d) OF THE
SECURITIES EXCHANGE ACT OF 1934
 
 
 
For the Fiscal Year Ended June 30, 2022
 
 
 
OR
 
 
☐
TRANSITION REPORT PURSUANT TO SECTION 13 OR 15(d) OF
THE SECURIT


In [17]:
embedding_function = SentenceTransformerEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")



/tmp/ipython-input-3551722535.py:1: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_function = SentenceTransformerEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")
/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access p

In [18]:
# Create a Chroma vector store and persist it locally
db = Chroma.from_documents(documents=docs, embedding=embedding_function, persist_directory="chroma_store")

# Save to disk
db.persist()

/tmp/ipython-input-2790807997.py:5: LangChainDeprecationWarning: Since Chroma 0.4.x the manual persistence method is no longer supported as docs are automatically persisted.
  db.persist()


Set {"k": 3} which means the retriever picks up 3 chunks or documents

In [19]:
persist_directory = "chroma_store"

embedding_model = SentenceTransformerEmbeddings(
    model_name="sentence-transformers/all-MiniLM-L6-v2"
)

vector_store = Chroma(
    persist_directory=persist_directory,
    embedding_function=embedding_model
)

retriever = vector_store.as_retriever(search_kwargs={"k": 3})

/tmp/ipython-input-3897217600.py:7: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-chroma package and should be used instead. To use it run `pip install -U :class:`~langchain-chroma` and import as `from :class:`~langchain_chroma import Chroma``.
  vector_store = Chroma(


QA chain is defined where chain type is set as 'stuff' which is a simple mode whch will stuff all the retrived chunks into a single prompt

In [20]:
qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=retriever,
    return_source_documents=True
)

In [21]:
query = "What were the company's total revenues for the fiscal year that ended on June 30, 2022?"
result = qa_chain.invoke(query)


In [22]:
result['result']

"The company's total revenues for the fiscal year that ended on June 30, 2022, were $198,270 million."

By reviwing the source document we can see the value **$198,270** there hence we know the answer is correct

In [31]:
result['source_documents'][0].page_content

'services, and customer service and support. Each\nallocation is measured differently based on the\nspecific facts and circumstances of the costs being\nallocated.\nSegment revenue and operating income were as follows\nduring the periods presented:\n \n(In millions)\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \nYear Ended June 30,\n \n \n2022\n \n \n \n2021\n \n \n \n2020\n \n \n \n \n \n \n \n \n \n \n \n \n \n \nRevenue\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \nProductivity and Business Processes\n \n$\n63,364\n \n \n$\n53,915\n \n \n$\n46,398\n \nIntelligent Cloud\n \n \n75,251\n \n \n \n60,080\n \n \n \n48,366\n \nMore Personal Computing\n \n \n59,655\n \n \n \n54,093\n \n \n \n48,251\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \nTotal\n \n$\n198,270\n \n \n$\n168,088\n \n \n$\n143,015\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \nOperating Income\n \n \n \n \n \n \n \n \n \n \n \n \